sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,591 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Manages update artifacts stored on this sled. The implementation is a very
+//! basic content-addressed object store.
+//!
+//! GET operations are handled by the "Repo Depot" API, which is deliberately
+//! a separate Dropshot service from the rest of Sled Agent. This is to avoid a
+//! circular logical dependency, because we expect Sled Agent to fetch artifacts
+//! it does not have from another Repo Depot that does have them (at Nexus's
+//! direction). This API's implementation is also part of this module.
+//!
+//! POST, PUT, and DELETE operations are handled by the Sled Agent API.
+
+use std::io::ErrorKind;
+use std::net::SocketAddrV6;
+use std::sync::LazyLock;
+use std::time::Duration;
+
+use camino::{Utf8Path, Utf8PathBuf};
+use camino_tempfile::NamedUtf8TempFile;
+use dropshot::{
+    Body, ConfigDropshot, FreeformBody, HttpError, HttpResponseOk,
+    HttpServerStarter, Path, RequestContext, StreamingBody,
+};
+use futures::{Stream, TryStreamExt};
+use illumos_utils::zpool::ZPOOL_MOUNTPOINT_ROOT;
+use omicron_common::address::REPO_DEPOT_PORT;
+use omicron_common::disk::{DatasetKind, DatasetsConfig};
+use omicron_common::update::ArtifactHash;
+use repo_depot_api::*;
+use sha2::{Digest, Sha256};
+use sled_storage::manager::StorageHandle;
+use slog::{error, info, Logger};
+use tokio::fs::File;
+use tokio::io::AsyncWriteExt;
+
+const TEMP_SUBDIR: &str = "tmp";
+
+#[derive(Clone)]
+pub(crate) struct ArtifactStore<T: StorageBackend = StorageHandle> {
+    log: Logger,
+    storage: T,
+}
+
+impl<T: StorageBackend> ArtifactStore<T> {
+    pub(crate) fn new(log: &Logger, storage: T) -> ArtifactStore<T> {
+        ArtifactStore {
+            log: log.new(slog::o!("component" => "ArtifactStore")),
+            storage,
+        }
+    }
+}
+
+impl ArtifactStore {
+    pub(crate) async fn start(
```

Comment:
I don't follow why this is specific to the real sled agent.  Is there any downside to executing this code for the simulated and test ones, too?  It seems like it just won't find anything to clean up.

Reply:
It would also start an HTTP server for the repo depot API on a fixed port, which would fail in parallel tests. The cleanup code is only in `start` because `new` isn't already async nor fallible.

Reply:
... I should also clarify the comment to this effect, but by making this start function specific over `ArtifactStore<StorageHandle>`, I'm avoiding wrapping the `sled_storage::error::Error` in `sled_agent::artifact_storage::Error`, so that I can use it directly in `sled_agent::artifact_storage::StartError`. This is somewhat arbitrary but it feels nicer to me.